snuba/query/matcher.py

```diff
@@ -0,0 +1,201 @@
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+from dataclasses import dataclass, field
+from datetime import date, datetime
+from typing import Generic, Mapping, Optional, Tuple, TypeVar
+
+from snuba.query.expressions import Column as ColumnExpr
+from snuba.query.expressions import Expression, FunctionCall
+
+TNode = TypeVar("TNode")
+
+
+@dataclass(frozen=True)
+class MatchResult:
+    """
+    Contains the parameters found in an expression by a Pattern.
+    Parameters are the equivalent to groups in regular expressions, they
+    are named parts of the expression we are matching that are identified
+    when a valid match is found.
+    """
+
+    expressions: Mapping[str, Expression] = field(default_factory=dict)
+    strings: Mapping[str, Optional[str]] = field(default_factory=dict)
+    ints: Mapping[str, Optional[int]] = field(default_factory=dict)
+    floats: Mapping[str, Optional[float]] = field(default_factory=dict)
+    dates: Mapping[str, Optional[date]] = field(default_factory=dict)
+    datetimes: Mapping[str, Optional[datetime]] = field(default_factory=dict)
+
+    def merge(self, values: MatchResult) -> MatchResult:
+        return MatchResult(
+            expressions={**self.expressions, **values.expressions},
+            strings={**self.strings, **values.strings},
+            ints={**self.ints, **values.ints},
+            floats={**self.floats, **values.floats},
+            dates={**self.dates, **values.dates},
+            datetimes={**self.datetimes, **values.datetimes},
+        )
+
+
+class Pattern(ABC, Generic[TNode]):
+    """
+    Tries to match a given node (like an AST Expression) with the rules provided
+    when instantiating this class. This is meant to work like a regular expression
+    but for AST expressions.
+
+    When a node is provided to this method, if this method return an instance of
+    MatchResult it means the node fully matches the pattern.
+    If the pattern defines named parameters, their value is returned as part of
+    the MatchResult object returned.
+    More on how to define parameters in the Param class.
+
+    This is not supposed to ever throw as long as the node is properly is a valid
+    data structure, even if it does not match the pattern.
+    """
+
+    @abstractmethod
+    def match(self, node: TNode) -> Optional[MatchResult]:
```

Comment:
Making this generic provides one advantage. Mypy can apply a few static checks on the validity of the Pattern provided by the user. For example mypy will not accept a pattern like this:
```
Function(None, Column(...),  None)
```
Because the function name is required to be an Optional[Pattern[str]], so. you cannot try to match a column where the function name is expected, thus making it harder to write impossible Pattern expressions.